File: common/src/main/java/com/ssblur/scriptor/word/action/RainAction.java
package com.ssblur.scriptor.word.action;

import com.ssblur.scriptor.api.word.Action;
import com.ssblur.scriptor.api.word.Descriptor;
import com.ssblur.scriptor.helpers.targetable.EntityTargetable;
import com.ssblur.scriptor.helpers.targetable.Targetable;
import com.ssblur.scriptor.api.word.descriptor.StrengthDescriptor;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.LightningBolt;

public class RainAction extends Action {
  @Override
  public void apply(Targetable caster, Targetable targetable, Descriptor[] descriptors) {
    if(targetable.getLevel().isClientSide) return;
    double strength = 2;
    for(var d: descriptors) {
      if(d instanceof StrengthDescriptor strengthDescriptor)
        strength += strengthDescriptor.strengthModifier();
    }

    var serverLevel = (ServerLevel) targetable.getLevel();
    serverLevel.setWeatherParameters(0, (int) strength * 2000, true, strength > 6);

    strength -= 9;
    for(double i = strength; i >= 0; i -= 5) {
      double x = Math.random() * 12 - 6;
      double y = Math.random() * -1;
      double z = Math.random() * 12 - 6;
      x += Math.signum(x) * 2;
      z += Math.signum(z) * 2;
      LightningBolt bolt = new LightningBolt(EntityType.LIGHTNING_BOLT, serverLevel);
      if(caster instanceof EntityTargetable entityTargetable && entityTargetable.getTargetEntity() instanceof ServerPlayer player)
        bolt.setCause(player);
      bolt.setPos(targetable.getTargetPos().add(x, y, z));
      serverLevel.addFreshEntity(bolt);
    }
  }
  @Override
  public Cost cost() { return new Cost(51, COSTTYPE.ADDITIVE); }
}

File: common/src/main/kotlin/com/ssblur/scriptor/word/action/RainAction.kt
package com.ssblur.scriptor.word.action

import com.ssblur.scriptor.api.word.Action
import com.ssblur.scriptor.api.word.Descriptor
import com.ssblur.scriptor.helpers.targetable.EntityTargetable
import com.ssblur.scriptor.helpers.targetable.Targetable
import com.ssblur.scriptor.word.descriptor.power.StrengthDescriptor
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.LightningBolt
import kotlin.math.sign

class RainAction : Action() {
    override fun apply(caster: Targetable, targetable: Targetable, descriptors: Array<Descriptor>) {
        if (targetable.level.isClientSide) return
        var strength = 2.0
        for (d in descriptors) {
            if (d is StrengthDescriptor) strength += d.strengthModifier()
        }

        val serverLevel = targetable.level as ServerLevel
        serverLevel.setWeatherParameters(0, strength.toInt() * 2000, true, strength > 6)

        strength -= 9.0
        var i = strength
        while (i >= 0) {
            var x = Math.random() * 12 - 6
            val y = Math.random() * -1
            var z = Math.random() * 12 - 6
            x += sign(x) * 2
            z += sign(z) * 2
            val bolt = LightningBolt(EntityType.LIGHTNING_BOLT, serverLevel)
            if (caster is EntityTargetable && caster.targetEntity is ServerPlayer) bolt.cause = (caster.targetEntity as ServerPlayer)
            bolt.setPos(targetable.targetPos.add(x, y, z))
            serverLevel.addFreshEntity(bolt)
            i -= 5.0
        }
    }

    override fun cost() = Cost(51.0, COSTTYPE.ADDITIVE)
}
